### PR TITLE
Adds 20 new and exciting pill colors!

### PR DESCRIPTION
**Pull request by Goldsteel**
All credit for the sprites goes to the thread anon.

Adds 20 new pills to the chem master.

<details>

image

With many chems, the bottom buttons are concealed such that you will need to resize the window or scroll down when doing 4-5 chems at a time.

image

This may cause some negative responses from people, but I don't know if we can resize this window by default or make some UI changes to compensate. Many machines already need you to scroll through lots of options like the protolathe and biogenerator. Scrolling is only really needed with more than a few chems, so you won't be inconvenienced too much.

image

They look fairly nice in game. You can now make "FUCKING DEATH" floor pills in black or perhaps make a pill called "LEMSIP" in the lemsip colours.

-------------- UPDATE -------------------

image

Now we have a toggle pill colours button for all your UI needs. No more window resizing.

-------------- UPDATE 2.0 -------------------

image

You no longer have to resize the window when dealing with tons of chems. Additional pills are now optional and can be toggled off so you don't have to see them.

image

</details>

-------------- UPDATE 3.0 -------------------

image

This is what the final edition looks like. Hitting the button toggles the pill colours between old and new modes, with or without a beaker added.

:cl:
 * rscadd: Adds 20 new pill colours to the chem master. Toggle them on and off at your leisure.

### CODE REVIEW

**Comment by boy2mantwicethefam** [comment]
Louis

**Comment by strawberrytan** [comment]
gotta grab everything i can

**Comment by DrSnips** [comment]
I would just make the window larger while you're at it.

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
Expand button feels clumsy, I'll never want to remove ALL pill icons, especially since the UI is already made to fit 2 rows snugly almost all the time. ~~Plus it's ugly positioned~~
I'd rather have the expand/contract still show the default 2 rows when contracted

**Comment by Goldsteel** [comment]
I'll see if I can make the window bigger by default. It should be easy enough to have a "show more pills" button and hide the extended range. Where exactly should the button for pill hiding be?

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
My suggestion is add an argument to generate_pill_icon_div() saying how many pills should be shown. Expand/contract should be in the same div as the pill icon themselves ideally

**Comment by Goldsteel** [comment]
@I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE There you go my friend

File: code/modules/reagents/machinery/chem_master.dm [hunk 1]
```diff
@@ -433,15 +441,26 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
 						[pill_icon_cache[i]]
 					</div>
 				</a>"}
 		if (i%10 == 0)
-			dat +="<br>"
-	dat += "</div>"
+			dat += "<br>"
+
+	if (iconToggle)
+		for(var/i = (MAX_PILL_SPRITE/2)+1 to MAX_PILL_SPRITE) // show the second half, rounding needed for odd
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: yikes
The functionality is good but this is really bad copypaste
Try adding a new argument to `generate_pill_icon_div()`, then modify the line inside that proc where it says `for(var/i = 1 to MAX_PILL_SPRITE)` and replace `MAX_PILL_SPRITE` with your new argument
Your final proc should look something like:
`/obj/machinery/chem_master/proc/generate_pill_icon_div(var/the_number_up_to_which_we_want_to_draw_pill_icons)
	var/dat = list()
	dat += "<HR>"
	dat += "<div class='pillIconsContainer'>"
	for(var/i = 1 to the_number_up_to_which_we_want_to_draw_pill_icons)
		dat += {"<a href="?src=\ref[src]&pill_sprite=[i]" class="pillIconWrapper[i == text2num(pillsprite) ? " linkOnMinimal" : ""]">
					<div class="pillIcon">
						[pill_icon_cache[i]]
					</div>
				</a>"}
		if (i%10 == 0)
			dat +="<br>"
	dat += "</div>"
	return dat`
Then when you call `generate_pill_icon_div()` you can give it an argument of 20 or 40 depending on what the setting is set to

File: code/modules/reagents/machinery/chem_master.dm [hunk 2]
```diff
@@ -452,14 +471,14 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
-	if(!beaker)
+	if(!beaker )
 		dat += "Please insert a beaker.<BR>"
 		if(!condi)
 			if(src.loaded_pill_bottle)
 				dat += "<A href='?src=\ref[src];ejectp=1'>Eject Pill Bottle \[[loaded_pill_bottle.contents.len]/[loaded_pill_bottle.storage_slots]\]</A><BR><BR>"
 			else
 				dat += "No pill bottle inserted.<BR><BR>"
-			dat += generate_pill_icon_div()
+			//dat += generate_pill_icon_div()
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: This is intended so you can set the pill icon while no beaker is inserted to recolor timerelease pills and such

File: code/modules/reagents/machinery/chem_master.dm [hunk 3]
```diff
@@ -582,8 +601,14 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 			//
 			// PILL ICONS
 			//
+
+			dat += "<HR><a href='?src=\ref[src];pill_icon_toggle=1'>Toggle Additional Pill Icons</a>"
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: You're adding an unnecessary horizontal line, try adding this inside the generate_pill_icon_div() proc

File: code/modules/reagents/machinery/chem_master.dm (hunk 1 above)
Comment (Goldsteel) [reply code/modules/reagents/machinery/chem_master.dm]: already working on it

File: code/modules/reagents/machinery/chem_master.dm (hunk 2 above)
Comment (Goldsteel) [reply code/modules/reagents/machinery/chem_master.dm]: readded

File: code/modules/reagents/machinery/chem_master.dm [hunk 4]
```diff
@@ -595,7 +620,7 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 					<A href='?src=\ref[src];createbottle_multiple=1;createempty=1'>Create empty bottles</A><BR>"}
 
 	dat = jointext(dat,"")
-	var/datum/browser/popup = new(user, "[windowtype]", "[name]", 475, 500, src)
+	var/datum/browser/popup = new(user, "[windowtype]", "[name]", 500, 800, src)
```
[the remark this answers is not in the document]
Comment (Goldsteel) [reply code/modules/reagents/machinery/chem_master.dm]: easy to change but basically it was because people dont want to resize the window much

**Comment by Goldsteel** [comment]
@I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE Hopefully that makes things a bit less crap, but the only problem now is that you can't toggle the pills without putting a beaker in first. I don't know if that's a problem people care about.

File: code/modules/reagents/machinery/chem_master.dm [hunk 5]
```diff
@@ -31,6 +33,8 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 	var/max_bottle_size = 30
 	var/max_pill_count = 20
 	var/max_pill_size = 50
+	var/iconToggle = 0
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: Delete this variable and just use `pill_display_number` instead

File: code/modules/reagents/machinery/chem_master.dm [hunk 6]
```diff
@@ -295,6 +299,11 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 			src.updateUsrDialog()
 			return 1
 
+		else if(href_list["pill_icon_toggle"])
+			iconToggle = !iconToggle
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: Update `pill_display_number` here not every time the UI is rendered

File: code/modules/reagents/machinery/chem_master.dm [hunk 7]
```diff
@@ -429,19 +438,20 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 	for(var/i = 1 to MAX_PILL_SPRITE)
 		pill_icon_cache += bicon(icon('icons/obj/chemical.dmi', "pill" + num2text(i)))
 
-/obj/machinery/chem_master/proc/generate_pill_icon_div()
+/obj/machinery/chem_master/proc/generate_pill_icon_div(pill_display_number)
 	var/dat = list()
 	dat += "<HR>"
 	dat += "<div class='pillIconsContainer'>"
-	for(var/i = 1 to MAX_PILL_SPRITE)
+	for(var/i = 1 to pill_display_number) // show the first half, rounding needed for odd
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: Leftover comment

File: code/modules/reagents/machinery/chem_master.dm [hunk 8]
```diff
@@ -452,14 +462,14 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 	user.set_machine(src)
 
 	var/dat = list()
-	if(!beaker)
+	if(!beaker )
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: Trailing space

File: code/modules/reagents/machinery/chem_master.dm [hunk 9]
```diff
@@ -582,7 +592,18 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 			//
 			// PILL ICONS
 			//
-			dat += generate_pill_icon_div()
+
+			dat += "<a href='?src=\ref[src];pill_icon_toggle=1'>Toggle Additional Pill Icons</a>"
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: This button will show even if no beaker is added if you just put it into the `generate_pill_icon_div()` proc like I said

File: code/modules/reagents/machinery/chem_master.dm [hunk 10]
```diff
@@ -429,18 +440,21 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 	for(var/i = 1 to MAX_PILL_SPRITE)
 		pill_icon_cache += bicon(icon('icons/obj/chemical.dmi', "pill" + num2text(i)))
 
-/obj/machinery/chem_master/proc/generate_pill_icon_div()
+/obj/machinery/chem_master/proc/generate_pill_icon_div(pill_display_number)
 	var/dat = list()
 	dat += "<HR>"
+	dat += "<a href='?src=\ref[src];pill_icon_toggle=1'>Toggle Additional Pill Icons</a>"
+	dat += "<HR>"
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: You've got an extra <HR> here

File: code/modules/reagents/machinery/chem_master.dm [hunk 11]
```diff
@@ -452,14 +466,16 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
...
+
 	if(!beaker)
 		dat += "Please insert a beaker.<BR>"
 		if(!condi)
 			if(src.loaded_pill_bottle)
 				dat += "<A href='?src=\ref[src];ejectp=1'>Eject Pill Bottle \[[loaded_pill_bottle.contents.len]/[loaded_pill_bottle.storage_slots]\]</A><BR><BR>"
 			else
 				dat += "No pill bottle inserted.<BR><BR>"
-			dat += generate_pill_icon_div()
+		dat += generate_pill_icon_div(pill_display_number)
```
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [inline code/modules/reagents/machinery/chem_master.dm]: You changed the tabulation here

Comment (Goldsteel) [reply code/modules/reagents/machinery/chem_master.dm]: fixed

File: code/modules/reagents/machinery/chem_master.dm (hunk 10 above)
Comment (Goldsteel) [reply code/modules/reagents/machinery/chem_master.dm]: fixed

File: code/modules/reagents/machinery/chem_master.dm (hunk 11 above)
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [reply code/modules/reagents/machinery/chem_master.dm]: Did you forget to commit this fix?

File: code/modules/reagents/machinery/chem_master.dm (hunk 10 above)
Comment (I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE) [reply code/modules/reagents/machinery/chem_master.dm]: Did you forget to commit this fix?

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
:ok_hand:
Hopefully I didn't noggin jog you too hard with the code but it's great now
Update the OP

File: code/modules/reagents/machinery/chem_master.dm [hunk 12]
```diff
@@ -429,18 +440,20 @@ var/global/list/pillIcon2Name = list("oblong purple-pink", "oblong green-white",
 	for(var/i = 1 to MAX_PILL_SPRITE)
 		pill_icon_cache += bicon(icon('icons/obj/chemical.dmi', "pill" + num2text(i)))
 
-/obj/machinery/chem_master/proc/generate_pill_icon_div()
+/obj/machinery/chem_master/proc/generate_pill_icon_div(pill_display_number)
 	var/dat = list()
 	dat += "<HR>"
+	dat += "<a href='?src=\ref[src];pill_icon_toggle=1'>Toggle Additional Pill Icons</a><br><br>"
```
[the remark this answers is not in the document]
Comment (Goldsteel) [reply code/modules/reagents/machinery/chem_master.dm]: done

**Review by MadmanMartian — APPROVED** [review]

**Pull request merged**